OptimoveNotificationServiceExtension/Sources/CategoryManager.swift

```diff
@@ -1,92 +1,55 @@
+        static let MAX_DYNAMIC_CATEGORIES = 128
+        static let DYNAMIC_CATEGORY = OptimobileUserDefaultsKey.DYNAMIC_CATEGORY.rawValue
     }
 
-    static func getCategoryIdForMessageId(messageId: Int) -> String {
-        return String(format: DYNAMIC_CATEGORY_IDENTIFIER, messageId)
+    static func getCategoryId(messageId: Int) -> String {
+        return "__kumulos_category_\(messageId)__"
     }
 
-    static func registerCategory(category: UNNotificationCategory) {
-        var categorySet = sharedInstance.getExistingCategories()
-        var storedDynamicCategories = sharedInstance.getExistingDynamicCategoriesList()
+    static func registerCategory(_ category: UNNotificationCategory) async {
+        var systemCategories = await UNUserNotificationCenter.current().notificationCategories()
+        var storedCategoryIds = readCategoryIds()
+
+        systemCategories.insert(category)
+        storedCategoryIds.insert(category.identifier)
 
-        categorySet.insert(category)
-        storedDynamicCategories.append(category.identifier)
+        let (categories, categoryIds) = maybePruneCategories(categories: systemCategories, categoryIds: storedCategoryIds)
 
-        sharedInstance.pruneCategoriesAndSave(categories: categorySet, dynamicCategories: storedDynamicCategories)
+        UNUserNotificationCenter.current().setNotificationCategories(categories)
+        writeCategoryIds(categoryIds)
 
         // Force a reload of the categories
-        _ = sharedInstance.getExistingCategories()
+        await UNUserNotificationCenter.current().notificationCategories()
     }
 
-    private func getExistingCategories() -> Set<UNNotificationCategory> {
-        var returnedCategories = Set<UNNotificationCategory>()
-
-        UNUserNotificationCenter.current().getNotificationCategories { (categories: Set<UNNotificationCategory>) in
-            returnedCategories = Set<UNNotificationCategory>(categories)
-
-            self.categoryReadLock.signal()
-        }
-
-        _ = categoryReadLock.wait(timeout: DispatchTime.now() + DispatchTimeInterval.seconds(5))
-
-        return returnedCategories
+    static func readCategoryIds() -> Set<String> {
+        let array = UserDefaults.standard.object(forKey: Constants.DYNAMIC_CATEGORY) as? [String] ?? []
+        return Set(array)
     }
 
-    private func getExistingDynamicCategoriesList() -> [String] {
-        dynamicCategoryLock.wait()
-        defer {
-            dynamicCategoryLock.signal()
-        }
-
-        if let existingArray = UserDefaults.standard.object(forKey: OptimobileUserDefaultsKey.DYNAMIC_CATEGORY.rawValue) {
-            return existingArray as! [String]
-        }
-
-        let newArray = [String]()
-
-        UserDefaults.standard.set(newArray, forKey: OptimobileUserDefaultsKey.DYNAMIC_CATEGORY.rawValue)
-
-        return newArray
+    static func writeCategoryIds(_ ids: Set<String>) {
+        UserDefaults.standard.set(Array(ids), forKey: Constants.DYNAMIC_CATEGORY)
```

Comment:
The standard iOS persistence will be replaced with the storage obstruction in this bulk of PRs